src/Adapter/MSTest.TestAdapter/Execution/TestMethodRunner.cs

```diff
@@ -370,16 +366,39 @@ private TestResult[] ExecuteTestWithDataSource(UTF.ITestDataSource? testDataSour
         string? displayName = StringEx.IsNullOrWhiteSpace(_test.DisplayName)
             ? _test.Name
             : _test.DisplayName;
-        if (testDataSource != null)
+
+        string? ignoreFromTestDataRow = null;
+        string? displayNameFromTestDataRow = null;
+        if (data?.Length == 1 && data[0] is ITestDataRow testDataRow)
+        {
+            object? dataFromTestDataRow = testDataRow.Value;
+            ignoreFromTestDataRow = testDataRow.IgnoreMessage;
+            displayNameFromTestDataRow = testDataRow.DisplayName;
+
+            data = TestDataSourceHelpers.TryHandleTupleDataSource(dataFromTestDataRow, _testMethodInfo.ParameterTypes, out object?[] tupleExpandedToArray)
+                ? tupleExpandedToArray
+                : [dataFromTestDataRow];
+        }
```

Comment:
This block is exactly the same as in the discovery, it could be useful to have a static method for it with either multiple out or a special type to be returned.